dnscache.go

```diff
@@ -64,6 +66,22 @@ func (r *Resolver) Refresh(clearUnused bool) {
 	}
 }
 
+// Dial is a function to use in DialContext for http.Transport.
+func (r *Resolver) Dial(ctx context.Context, network string, addr string) (conn net.Conn, err error) {
```

Comment:
This is a simplistic example implementation of a dial with custom resolution. It misses a lot of features like dual stack support, correct handling of context tracing, randomization of adds etc. This implementation will also choke on "unix", "unixgram" and "unixpacket" network.

Reply:
Keep it an example. Semantically, a `dnscache.Dial` would be a way to dial the DNS cache, not dial an arbitrary network location using the dnscache.

Reply:
but this will invalidate the pull request and it's purpose.
I'll think what can I do to address this issue (but basically stuck to more docs and renaming options for now).

Reply:
The main issue is that `net.Dialer.Resolver` field does only accept a concrete type instead of an interface. Otherwise we could have just implemented the `net.Resolver` interface in order to provide a caching layer to any `net.Dialer` and everything would have been transparent.

It would be still possible to trick the `net.Resolver` type by implementing a `Dial` method for this type that would decode DNS frames and return cached entries there, and forcing the `PreferGo` to `true`. It would be complex, waste CPU cycles and generate GC garbage for nothing though.

Note that I never used this lib myself, it was just a quick implementation for a friend. That's why it didn't get much love. A more appropriate approach to DNS caching is to install unbound on the system instead of relying on Go, but that's not always possible/easy, especially in a containerized environment (hence this lib).

Reply:
actually I have an application in that it's useful. and works better that local unbound, because there are tens of thousands DNS request per every second.

the alternative was... well, rewriting it in Java that has DNS cache built-in.

Reply:
(and yes, all my requests are duplicate. the application is an HTTP stress tester)

Reply:
and yes, I'm aware of fasthttp library that advertises DNS caching. with my concurrency it just does not work. but the dnscache does (I actually had to pre-resolve all the hostnames manually for best results, I see no straightforward way of doing it in fasthttp).

Reply:
That's a great use-case. You seem to say that not having the dial method packaged with the lib is a deal breaker. Why don't you have an implementation of the dial in your code? You will get more control this way.

Reply:
I actually don't like specifying this function inline in http.Transport

Unfortunately, I cannot add a method to 3-rd party library in my code.
not I can declare it in my package, as it relies to have a dnscache.Resolver object initialized outside of it's scope (and cannot change the function signature).

in fact, I'd prefer to manage my own fork if you won't take this change.
but I'll try.

Reply:
rebased my branch, documented all the concerns in the docs to the Dial function.

actually I think that if somebody will need more features in this implementation, he can contribute them to this library, so everyone will benefit from having even this minimalistic implementation in library.